### PR TITLE
Implement functionality to create anchored time-index

### LINKED ISSUE #8: Implementing the TimeIndex class - the constructor method
In #7 we drafted a preliminary interface for the main time-index object. Now, we want to start implementing it, starting with the constructor (the `__init__` method). 

Also see the discussion in https://github.com/AI4S2S/proto/issues/3

### PR DESCRIPTION
**Pull request by Peter9192**
Far from ready, but good enough to build upon. 

closes #8 

How to use:

```py
import s2s.time

cal = s2s.time.AdventCalendar(anchor_date=(11, 30), freq='7d')
cal
>> "AdventCalendar(month=11, day=30, freq=7d, n=52)"

print(cal)
>> "52 periods of 7d leading up to 11/30."

# get time indexers
year = cal.map_year(2020)
years = cal.map_years(2020, 2022)
years_flat = cal.map_years(2020, 2022, flat=True)
```

### CODE REVIEW

File: tests/test_time.py [hunk 1]
```diff
@@ -3,78 +3,108 @@
...
-            index.discard(max_lag=5)
+            cal.discard(max_lag=5)
+
+    def test_map_year(self):
+        cal = AdventCalendar(anchor_date=(12, 31), freq="180d")
+        year = cal.map_year(2020)
+        expected = np.array(
+            [interval("2020-07-04", "2020-12-31"), interval("2020-01-06", "2020-07-04")]
+        )
+        assert np.array_equal(year.values, expected)
```
Comment (geek-yang) [inline tests/test_time.py:39]:
```suggestion
        assert np.array_equal(year, expected)
```

Comment (geek-yang) [inline tests/test_time.py:39]:
```suggestion
assert np.array_equal(year, expected)
```

**Review by geek-yang — APPROVED** [review]
@Peter9192 Fantastic work!! Thanks for the breach! This really helps shape the following work. I only have minor comments regarding the doc strings, `mypy` constraint and some tiny things related to tests. We can also implement these changes later if you don't have time for now. Feel free to merge it.

File: s2s/time.py [hunk 2]
```diff
@@ -11,24 +11,46 @@
...
 
-class TimeIndex:
-    """TimeIndex anchored to a date or period of interest."""
+class AdventCalendar:
+    """Countdown time to anticipated anchor date or period of interest."""
 
-    def __init__(self, anchor_date="2020-11-30", freq="7d", cycle_time="1yr"):
-        """Instantiate a basic index with minimal configuration.
+    def __init__(self, anchor_date=(11, 30), freq="7d"):
+        """Instantiate a basic calendar with minimal configuration.
```
[the remark this answers is not in the document]
Comment (Peter9192) [reply s2s/time.py:63]: Done. Look how nicely this renders: 

image

File: s2s/time.py [hunk 3]
```diff
@@ -11,24 +11,46 @@
...
 
 
-class TimeIndex:
-    """TimeIndex anchored to a date or period of interest."""
+class AdventCalendar:
+    """Countdown time to anticipated anchor date or period of interest."""
 
-    def __init__(self, anchor_date="2020-11-30", freq="7d", cycle_time="1yr"):
-        """Instantiate a basic index with minimal configuration.
+    def __init__(self, anchor_date=(11, 30), freq="7d"):
```
[the remark this answers is not in the document]
Comment (Peter9192) [reply s2s/time.py]: I started adding type annotations, but we need to properly setup Mypy. I created #12 for that.

**Comment by Peter9192** [comment]
Ready for another look @geek-yang

File: s2s/time.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (geek-yang) [reply s2s/time.py:63]: Looks awesome!

**Review by geek-yang — APPROVED** [review]
@Peter9192 Thanks for addressing the comments and this looks really awesome! Just go ahead and merge it!

Also, thanks for creating issue #15 . I will try to set it up quickly as then we can always keep all the docs related things on track during development.

**Pull request merged**